test/mock/ws_server.go

```diff
@@ -88,38 +123,71 @@ func (c *client) send() {
 	}
 }
 
-var internalClientConnectChan = make(chan *client)
-var ClientConnectChan = make(chan *client, 1)
-var ClientDisconnectChan = make(chan *client)
-var ClientRmChan = make(chan *client, 5)
-var Clients = make(map[*client]*client)
+func NewClients() *clients {
+	return &clients{
+		list: make(map[*client]*client),
+	}
+}
 
-func DisconnectClient(c *client) {
-	c, ok := Clients[c]
-	if ok {
+type clients struct {
+	list map[*client]*client
+	sync.RWMutex
+}
+
+func (cl *clients) add(c *client) {
+	cl.Lock()
+	defer cl.Unlock()
+
+	cl.list[c] = c
+}
+
+func (cl *clients) del(c *client) {
+	cl.Lock()
+	defer cl.Unlock()
+
+	if _, ok := cl.list[c]; ok {
+		delete(cl.list, c)
 		close(c.SendChan)
+	}
+}
+
+func (cl *clients) get(c *client) *client {
+	cl.RLock()
+	defer cl.RUnlock()
+
+	c, ok := cl.list[c]
+	if ok {
+		return c
+	}
+	return nil
+}
+
+func (cl *clients) getAny() *client {
+	cl.RLock()
+	defer cl.RUnlock()
+
+	for _, c := range cl.list {
+		return c
+	}
+	return nil
+}
+
+func (cl *clients) Disconnect(c *client) {
+	c = cl.get(c)
+	if c != nil {
 		c.ws.Close()
-		//log.Printf("disconnect: %+v\n", c)
-		<-ClientDisconnectChan
+		<-c.disconnectChan
+		cl.del(c)
 	}
 }
 
-func run() {
+// Disconnect all clients.
+func (cl *clients) DisconnectAll() {
 	for {
-		select {
-		case c := <-internalClientConnectChan:
-			// todo: this is probably prone to deadlocks, not thread-safe
```

Comment:
This `WebsocketServer` was so broken that every time some tests were passing and some failing and never the same, and on go 1.8 they were actually always failing.